src/pudl_rmi/formatter_optimus.py

```diff
@@ -348,3 +354,69 @@ def add_balancing_account(
     logger.info(f"adding {len(balancing_records)} balancing account records.")
     ferc_deprish_eia = pd.concat([ferc_deprish_eia, balancing_records], join="outer")
     return ferc_deprish_eia
+
+
+def allocate_balancing_account_to_assets(
+    ferc_deprish_eia_w_ba: pd.DataFrame,
+    idk_ba: List[str] = [
+        "utility_name_eia",
+        "operational_status",
+        "ferc_acct_name",
+        "data_source",
+        "report_year",
+    ],
+    data_and_allocator_cols: dict = {
+        "plant_balance_w_common": ["plant_balance_w_common"],
+        "book_reserve_w_common": ["book_reserve_w_common"],
+        "depreciation_annual_epxns_w_common": ["depreciation_annual_epxns_w_common"],
+    },
+):
+    """Allocate the balancing account records across related assets.
+
+    We generate balancing account records to true up the depreciation studies
+    with the utility/FERC account level values reported to FERC1 via
+    :func:`add_balancing_account`. This function takes those balancing account
+    records and allocates them across assets. It utilizes
+    :func:`pudl_rmi.connect_deprish_to_ferc1._allocate_col`
+
+    Args:
+        ferc_deprish_eia_w_ba
+        idk_ba: list of ID columns to allocate across.
+        data_and_allocator_cols: dictionary of data columns to allocate (keys)
+            and lists of column(s) to allocate based on.
+
+    """
+    # split records
+    mask_ba = ferc_deprish_eia_w_ba.plant_name_eia == "Balancing_Account"
+    balancing_accounts = ferc_deprish_eia_w_ba.loc[mask_ba]
+    assets = ferc_deprish_eia_w_ba.loc[~mask_ba]
+    data_cols = list(data_and_allocator_cols.keys())
+    # merge the ba's onto the assets
+    assets_w_ba = pd.merge(
+        assets,
+        balancing_accounts[idk_ba + data_cols],
+        on=idk_ba,
+        suffixes=("", "_ba"),
+        how="left",
+        validate="m:1",
+    )
+    # allocate each col
+    for (data_col, allocator_cols) in data_and_allocator_cols.items():
```

Comment:
I could imagine refactoring `_allocate_col` to more quickly calculate all of the data columns all at once so we could  avoid this loop